### PR TITLE
allow retrying initial ssh connection

### PR DESCRIPTION
**Pull request by dontlaugh**

### CODE REVIEW

**Comment by dontlaugh** [comment]
@dcole-inviqa I needed a retry mechanism for my own infrastructure, too! I amended your commit with the recommendations from #1 

@bradrydzewski Let me know if there's anything else you'd like.

File: engine/engine_impl.go [hunk 1]
```diff
@@ -75,31 +76,18 @@ func (e *engine) Setup(ctx context.Context, spec *Spec) error {
...
-		logger.FromContext(ctx).
-			WithError(err).
-			WithField("hostname", spec.Server.Name).
-			WithField("user", spec.Server.User).
-			WithField("ip", instance.IP).
-			WithField("id", instance.ID).
-			Debug("failed to dial server")
 		return err
 	}
-	defer client.Close()
```
Comment (bradrydzewski) [inline engine/engine_impl.go:102]: any reason this is being removed?  I believe we still need to close the client connection once we are done.

**Comment by bradrydzewski** [comment]
we also have a more robust retry function here:
https://github.com/drone-runners/drone-runner-macstadium/blob/master/engine/engine.go#L373:L424

one important difference in the above function is that we close client connections if not nil before we retry. This prevents a potential fd leadk.

File: engine/engine_impl.go (hunk 1 above)
Comment (bradrydzewski) [reply engine/engine_impl.go:102]: this actually may be safer:

```
	client, err := e.retryDial(ctx, spec)
	if client != nil {
		defer client.Close()
	}
```

we use the above in our macstadium runner

Comment (dontlaugh) [reply engine/engine_impl.go:102]: This was probably a mistake. I'll update.

**Comment by dontlaugh** [comment]
Thanks for the review! I think I've addressed your comments.

I've followed your advice, but I'm curious about the file descriptor leak comment. It looks like it shouldn't happen from the code, as long as the underlying ssh library does what it's supposed to. Have you seen resources leak in the past? Does it show up on high-traffic servers?

**Pull request merged**